examples/advanced/nce_example.py

```diff
@@ -0,0 +1,237 @@
+    """
+        Generate a set of foil constraints
+        Pick m items and assign the opposite value.
+        If this results in an unfeasable assignment (surpassing the capacity), try again
+        @return A model
+    """
+    n = len(x)
+    foil_idx = np.random.choice(n,m, replace=False)
+    foil_vals = np.abs(1 - x[foil_idx])
+
+    if sum(foil_vals * weights[foil_idx]) > capacity:
+        if verbose:
+            print(f"\rGenerated unfeasable user query, retrying...({tries})", end="")
+        return generate_foil_knapsack(values, weights, capacity, x, m, tries+1)
+    else:
+        return extend_to_full_solution(values, 
+                                       weights, 
+                                       capacity, 
+                                       foil_idx,
+                                       foil_vals)\
+                ,foil_idx
+
+def extend_to_full_solution(values, weights, capacity, foil_idx, foil_vals):
+    """
+        Extend a given set of foil constraints to a full solution of the knapsack
+        Formally:
+            Given v and X, solve the COP 
+    """
+    xv = boolvar(shape=len(values), name="xv")
+    constraints = [xv[foil_idx] == foil_vals]
+    constraints += [sum(xv * weights) <= capacity]
+
+    model = Model(
+        constraints,
+        maximize= sum(xv * values)
+    )
+
+    if model.solve() is not False:
+        return xv.value()
+
+def master_problem(values, weights, capacity, x_d, foil_idx):
+    
+    """
+        Master problem: iteratively find better values for the c vector (new values are vector d)
+        Mapping of variable names to names in the paper (TODO: link papper)
+            - c = values
+            - d = d
+        @param x_d: A feasable solution found in the foil set X_v
+        @param foil_idx: A vector containing the items on which the user asked an explanation
+                         All other items must retain their original values
+    """
+    if verbose:
+        print(f"\n\n{'='*10} Solving the master problem {'='*10}")
+    known_solutions, i= [], 1
+    while i:
+        if verbose:
+            print(f"\nStarting iteration {i}")
+        d = intvar(0,INFINITY, values.shape, name="d")
+        x = boolvar(shape=len(x_d), name="x")
+        # The ususal knapsack constraint
+        constraints = [sum(x * weights) <= capacity]
+
+        # Extension of the knapsack problem to the master problem
+        # Ensure the newly found values vector results in a better solution than any known solutions
+        constraints += [sum(d * x) >= sum(d * known) for known in known_solutions]
+        # Ensure values are only modified at foil indices 
+        constraints += [d[i] == values[i] for i in range(len(values)) if i not in foil_idx]
+        # Ensure the foil values assigned by the user remain the same
+        constraints += [x[i] == bool(x_d[i]) for i in foil_idx]
+       
+        master_model = Model(
```

Comment:
you don't need to rebuild the 'master_model' from scratch inside the loop...

CPMpy supprots incremental solving,
so you can make the model once before the loop,

and then do master_model += [sum(d*x) >= sum(d*new_solution)]
which is a lot more elegant : )

Reply:
Made the master problem and sub problem iterative.
Thereby shortening the method massively.